### PR TITLE
validate backup name format before contacting the API server

### PR DESCRIPTION
**Pull request by samay43**
Fixes #<your issue number>
https://github.com/velero-io/velero/issues/10165

`velero backup create` currently doesn't validate the backup name locally
before contacting the Kubernetes API server. An invalid name (uppercase
letters, underscores, spaces, etc.) produces the same generic
"can't find Kubernetes config" or raw API rejection error as a valid one,
instead of a clear validation message.

This adds a check in Validate() using
k8s.io/apimachinery/pkg/util/validation.IsDNS1123Subdomain(), which is
already a project dependency, consistent with how other inputs (backup
type, flag combinations) are validated locally before hitting the cluster.

Tested locally with a kind cluster:
- Invalid names (e.g. "Invalid_Name!") now return a clear validation
  error immediately
- Valid names pass through unaffected and proceed to normal cluster
  operations

### CODE REVIEW

**Comment by kaovilai** [comment]
This validation is reached only after Complete() calls KubebuilderWatchClient(). Consequently, an invalid backup name still cannot be reported first when Kubernetes configuration/client initialization fails, which appears to be the behavior this PR intends to fix. Could the argument/name validation be performed through Cobra’s Args validator or otherwise before KubebuilderWatchClient() is called?

File: pkg/cmd/cli/backup/create.go [hunk 1]
```diff
@@ -191,9 +192,11 @@ func (o *CreateOptions) Validate(c *cobra.Command, args []string, f client.Facto
 		return err
 	}
 
-	// Ensure that unless FromSchedule is set, args contains a backup name
-	if o.FromSchedule == "" && len(args) != 1 {
-		return fmt.Errorf("a backup name is required, unless you are creating based on a schedule")
+	// Ensure the backup name is a valid Kubernetes resource name
+	if o.FromSchedule == "" {
```
Comment (adam-jian-zhang) [inline pkg/cmd/cli/backup/create.go]: Problem1:  We can not simply drop the condition len(args) != 1,  which validates required field "backup name"
Recommendation: Separate missing name validation from format validation

Problem2: user creates a backup based on a schedule and supplies a custom name (e.g. velero backup create INVALID_NAME! --from-schedule my-schedule), o.FromSchedule == "" evaluates to false. The PR's if o.FromSchedule == "" condition causes validation to be skipped entirely for custom schedule-based backup names.
Recommendation: Validating o.Name whenever o.Name != "" (regardless of --from-schedule) ensures all user-supplied backup names are checked.

**Comment by samay43** [comment]
Hi @kaovilai , @adam-jian-zhang  — thanks for catching both of these, you're right on both counts.

Moved the name-format check into the command's `Args `function so it runs before `Complete()/Validate()` — should be caught before any cluster contact happens now, not just when Complete() succeeds.

Also fixed the` --from-schedule `issue — the check now runs whenever a name is actually supplied, regardless of whether `--from-schedule `is set too. Tested locally and it's catching that case now:

```
$ velero backup create "Invalid_Name!" --from-schedule daily-backup
Error: invalid backup name "Invalid_Name!": a lowercase RFC 1123 subdomain must consist of...
```

Also noticed one of the existing tests was using a name (`nameToBeCreated`) that's actually invalid under DNS-1123 rules, which the stricter check picked up on. Fixed the test fixture rather than loosening the validation.

Tests are passing locally, let me know if I've missed anything.

File: pkg/cmd/cli/backup/create.go [hunk 2]
```diff
@@ -45,7 +46,17 @@ func NewCreateCommand(f client.Factory, use string) *cobra.Command {
 	c := &cobra.Command{
 		Use:   use + " NAME",
 		Short: "Create a backup",
-		Args:  cobra.MaximumNArgs(1),
+		Args: func(c *cobra.Command, args []string) error {
+			if err := cobra.MaximumNArgs(1)(c, args); err != nil {
+				return err
+			}
+			if len(args) == 1 {
```
Comment (adam-jian-zhang) [inline pkg/cmd/cli/backup/create.go:57]: We should add validation for backups from schedule as well here:
```golang
    fromSchedule, _ := c.Flags().GetString("from-schedule")
    if fromSchedule == "" && len(args) == 0 {
        return fmt.Errorf("a backup name is required, unless you are creating based on a schedule")
    }
    if len(args) == 1 {
    ....
```

File: pkg/cmd/cli/backup/create.go [hunk 3]
```diff
@@ -191,9 +202,11 @@ func (o *CreateOptions) Validate(c *cobra.Command, args []string, f client.Facto
 		return err
 	}
 
-	// Ensure that unless FromSchedule is set, args contains a backup name
-	if o.FromSchedule == "" && len(args) != 1 {
-		return fmt.Errorf("a backup name is required, unless you are creating based on a schedule")
+	// Ensure the backup name is a valid Kubernetes resource name
+	if o.FromSchedule == "" {
+		if errs := validation.IsDNS1123Subdomain(o.Name); len(errs) > 0 {
```
Comment (adam-jian-zhang) [inline pkg/cmd/cli/backup/create.go:215]: The schedule check and name check are separate check.
User could also provide custom backup name when creating from schedule also.
Current logic would skip check name where FromSchedule is set, and provided a invalid backup name.

So the validation logic should be:
```golang
	// Ensure that unless FromSchedule is set, a backup name is provided
	if o.FromSchedule == "" && o.Name == "" {
		return fmt.Errorf("a backup name is required, unless you are creating based on a schedule")
	}
	// Validate the backup name format whenever a name is provided
	if o.Name != "" {
		if errs := validation.IsDNS1123Subdomain(o.Name); len(errs) > 0 {
			return fmt.Errorf("invalid backup name %q: %s", o.Name, strings.Join(errs, "; "))
		}
	}
```

File: pkg/cmd/cli/backup/create.go (hunk 2 above)
Comment (samay43) [reply pkg/cmd/cli/backup/create.go:57]: Added the from-schedule-aware check here too, so CLI-level `Args `validation matches the `Options.Validate` logic — requires a name unless `--from-schedule `is set. Pushed in `52f7c2408`

File: pkg/cmd/cli/backup/create.go (hunk 3 above)
Comment (samay43) [reply pkg/cmd/cli/backup/create.go:215]: Thank you, the validation logic was skipping the name-format check whenever `FromSchedule` was set, even if a custom name was also provided. Updated it to separate the two checks: require a name only when `FromSchedule` isn't set, but validate the name format whenever a name is present. Pushed in `52f7c2408.`

**Comment by adam-jian-zhang** [comment]
@samay43 please fix lint

**Comment by samay43** [comment]
@adam-jian-zhang Fixed — indentation issue in the `Args` function, pushed in c27a661d1.

**Review by adam-jian-zhang — DISMISSED** [review]
LGTM

**Comment by kaovilai** [comment]
lint :)

**Comment by samay43** [comment]
@kaovilai lint test passed, thank you.

**Review by kaovilai — CHANGES_REQUESTED** [review]
Nice improvement for velero backup create NAME; it now matches API-server DNS1123-subdomain validation. However, the same failure mode still exists for schedule-derived names (<schedule>-YYYYMMDDHHMMSS). Please also validate generated names when --from-schedule omits NAME, and consider schedule-create validation that reserves 15 chars for the timestamp suffix (max schedule length 238 under current Kubernetes 253-char subdomain limit).

**Comment by samay43** [comment]
Thanks for catching this, @kaovilai you are right, I only handled the case where a name is explicitly passed in, and missed that  schedule-derived name could still run over the limit once the timestamp gets appended. I've added a check so that when `--from-schedule `is used without a `NAME`, the schedule name itself is validated with 15 characters reserved for the generated `-YYYYMMDDHHMMSS` suffix — so the final name can never exceed the 253-character DNS1123 limit. Added tests for both the boundary case (exactly at the limit) and just over it, to make sure it fails correctly.
added that in the efe1279 please look into it and make any suggestion if required, Thank you.

**Review by kaovilai — APPROVED** [review]

**Review by samay43 — COMMENTED** [review]
@adam-jian-zhang please look into it, thank you

**Review by adam-jian-zhang — APPROVED** [review]
LGTM

**Comment by samay43** [comment]
@blackpiglet please look into it , Thank You

**Comment by samay43** [comment]
@blackpiglet please look into it thank you.

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by samay43** [comment]
@shubham-pampattiwar thanks for the review. This is green across all checks and `mergeable_state` is clean, with approvals from you, @kaovilai and @adam-jian-zhang. Anything left you'd like me to change, or is it good to merge?
